File: code/Java/src/main/java/Infrastructure/SpaceComm/Message.java
package Infrastructure.SpaceComm;

import java.util.ArrayList;

import static java.lang.Integer.parseInt;

public class Message {

    private String x;
    private String y;
    private int commandsCount;
    private String direction;
    private String positionMessage;
    private String commandsMessage;

    public Message(ArrayList<String> datagrams) {
        this.positionMessage = parsePosition(datagrams);
        this.commandsMessage = parseCommands(datagrams);
    }

    private String parsePosition(ArrayList<String> datagrams) {
        for (String datagram :
                datagrams) {
            if (datagram.startsWith("X"))
                this.x = datagram.substring(1);
            if (datagram.startsWith("Y"))
                this.y = datagram.substring(1);
            if (datagram.startsWith("D"))
                this.direction = datagram.substring(1);
            if (datagram.startsWith("M")) {
                this.commandsCount = parseInt(datagram.substring(1));
            }
        }
        return String.format("100 100\n%s %s %s\n", this.x, this.y, this.direction);
    }

    @Override
    public String toString() {
        return String.format("%s%s", this.positionMessage, this.commandsMessage);
    }

    public boolean isValid() {
        return this.x != null &&
                this.y != null &&
                this.direction != null &&
                this.commandsCount > 0 &&
                this.commandsMessage.length() == this.commandsCount;
    }

    String parseCommands(ArrayList<String> datagrams) {
        String commandsMessage = "";
        for (int commandNumber = 1; commandNumber <= commandsCount; commandNumber++) {
            for (String datagram : datagrams) {
                if (datagram.startsWith(String.valueOf(commandNumber))) {
                    commandsMessage += datagram.substring(1);
                }
            }
        }
        return commandsMessage;
    }
}

File: code/CSharp/ConnascentMarsRover/Source/Infrastructure/SpaceComm/Message.cs
using System.Collections.Generic;

namespace Source.Infrastructure.SpaceComm
{
    public class Message
    {
        private string _x;
        private string _y;
        private int _commandsCount;
        private string _direction;
        private readonly string _positionMessage;
        private readonly string _commandsMessage;

        public Message(List<string> datagrams)
        {
            _positionMessage = ParsePosition(datagrams);
            _commandsMessage = ParseCommands(datagrams);
        }

        private string ParsePosition(List<string> datagrams)
        {
            foreach (var datagram in datagrams)
            {
                if (datagram.StartsWith("X"))
                    _x = datagram.Substring(1);
                if (datagram.StartsWith("Y"))
                    _y = datagram.Substring(1);
                if (datagram.StartsWith("D"))
                    _direction = datagram.Substring(1);
                if (datagram.StartsWith("M"))
                {
                    _commandsCount = int.Parse(datagram.Substring(1));
                }
            }

            return $"100 100\n{_x} {_y} {_direction}\n";
        }

        public override string ToString()
        {
            return $"{_positionMessage}{_commandsMessage}";
        }

        public bool IsValid()
        {
            return _x != null &&
                   _y != null &&
                   _direction != null &&
                   _commandsCount > 0 &&
                   _commandsMessage.Length == _commandsCount;
        }

        string ParseCommands(List<string> datagrams)
        {
            var commandsMessage = "";
            for (int commandNumber = 1; commandNumber <= _commandsCount; commandNumber++)
            {
                foreach (var datagram in datagrams)
                {
                    if (datagram.StartsWith(commandNumber.ToString()))
                    {
                        commandsMessage += datagram.Substring(1);
                    }
                }
            }

            return commandsMessage;
        }
    }
}